File: 프로그래머스/2/389479. 서버 증설 횟수/서버 증설 횟수.java
import java.util.*;

class Solution {
    public int solution(int[] players, int m, int k) {
        int answer = 0;
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for(int i=0;i<players.length;i++){
            while(!pq.isEmpty()){
                if(pq.peek()<i) pq.poll();
                else break;
            }
            int num = players[i];
            int server = num/m - pq.size();
            for(int j=0;j<server;j++){
                pq.add(i+k-1);
                answer++;
            }
        }
        return answer;
    }
}

File: 프로그래머스/2/389479. 서버 증설 횟수/서버 증설 횟수.js
function solution(players, m, k) {
    var answer = 0;
    const server = Array.from({length:24},()=>0)
    function server_create(index){
        for(let i=index;i<24&&i<index+k;i++){
            server[i] += m
        }
        answer++;
    }
    for(let i=0;i<24;i++){
        while(players[i]>=server[i]+m) server_create(i)
    }

    return answer;
}
